# Case Study on Probability for Data Science

# Problem Statement:

To make a suitable machine learning algorithm to predict if the mushroom is 
edible or poisonous (e or p) using the given dataset.
(Along with other ML algorithms, Naïve Bayes’ Classifier should be applied)
Also, if some data pre-processing is necessary do that as well.

# Attribute Information:

• cap-shape: bell=b, conical=c, convex=x, flat=f, knobbed=k, sunken=s 
• cap-surface: fibrous=f, grooves=g, scaly=y, smooth=s 
• cap-colour: brown=n, buff=b, cinnamon=c, Gray=g, green=r, pink=p, purple=u,
red=e, white=w, yellow=y 
• bruises: bruises=t, no=f 
• odour: almond=a, anise=l, creosote=c, fishy=y, foul=f, musty=m, none=n,
pungent=p, spicy=s 
• gill-attachment: attached=a, descending=d, free=f, notched=n 
• gill-spacing: close=c, crowded=w, distant=d 
• gill-size: broad=b, narrow=n 
• gill-colour: black=k, brown=n, buff=b, chocolate=h, grey=g, green=r, orange=o, 
pink=p, purple=u, red=e, white=w, yellow=y 
• stalk-shape: enlarging=e, tapering=t 
• Stalk-root: bulbous=b, club=c, cup=u, equal=e, rhizomorphs=z, rooted=r, 
missing=? 
• stalk-surface-above-ring: fibrous=f, scaly=y, silky=k, smooth=s 
• stalk-surface-below-ring: fibrous=f, scaly=y, silky=k, smooth=s 
• stalk-colour-above-ring: brown=n, buff=b, cinnamon=c, gray=g, orange=o,
pink=p, red=e, white=w, yellow=y 
• stalk-colour-below-ring: brown=n, buff=b, cinnamon=c, gray=g, orange=o,
pink=p, red=e, white=w, yellow=y 
• veil-type: partial=p, universal=u 
• veil-colour: brown=n, orange=o, white=w, yellow=y 
• ring-number: none=n, one=o, two=t 
• ring-type: cobwebby=c, evanescent=e, flaring=f, large=l, none=n, pendant=p,
sheathing=s, zone=z 
• spore-print-colour: black=k, brown=n, buff=b, chocolate=h, green=r, orange=o,
purple=u, white=w, yellow=y 
• population: abundant=a, clustered=c, numerous=n, scattered=s, several=v,
solitary=y 
• habitat: grasses=g, leaves=l, meadows=m, paths=p, urban=u, waste=w,
woods=d 
Please note the following
Use the mushroom data file attached along with the question.
Give headings to each step you are doing.
Do the case study in Python.
Create a repository in GitHub account as "Public".
Upload the notebook file (. ipynb) to the repository.
Please make sure that you are uploading the notebook file including the outputs as well.
Share the link of this notebook from GitHub in the online text editor provided in Paatshala.

# Importing Libraries

In [1]:
import pandas as pd
import numpy as np

# Loading Data

In [2]:
df=pd.read_csv('C:/Users/user/Downloads/mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [3]:
df.shape

(8124, 23)

In [4]:
df.columns

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

In [5]:
df.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [7]:
df.dtypes

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object

In [8]:
%whos

Variable   Type         Data/Info
---------------------------------
df         DataFrame         class cap-shape cap-<...>n[8124 rows x 23 columns]
np         module       <module 'numpy' from 'C:\<...>ges\\numpy\\__init__.py'>
pd         module       <module 'pandas' from 'C:<...>es\\pandas\\__init__.py'>


In [9]:
df.isna().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

insights
1.There are 8124 rows and 23 columns in our data set.
2.There are no null values.    
    

# splitting into feature set and class table

In [10]:
y=df.iloc[:,0]
x=df.iloc[:,1:]


In [11]:
x= pd.get_dummies(x)
x.head()

cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            1            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  cap-surface_y  \
0            1              0              0              1              0   
1            1              0              0              1              0   
2            0              0              0              1              0   
3            1              0              0              0              1   
4            1              0              0              1              0   

   ...  population_s  population_v  population_y  habitat_d  habitat_g  \
0  ...             1             0             0          0          0   
1  ...             0             0             0          0          1   
2  ...             0             0             0          0          0   
3  ...             1             0             0          0          0   
4  ...             0             0             0          0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 117 columns]

# splitting the data set into the training set and test set

In [45]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=42)

In [46]:
x_train.shape

(6093, 117)

In [47]:
y_train.shape

(6093,)

In [48]:
x_test.shape

(2031, 117)

In [49]:
y_test.shape

(2031,)

In [50]:
x.describe()

cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
count  8124.000000  8124.000000  8124.000000  8124.000000  8124.000000   
mean      0.055638     0.000492     0.387986     0.101920     0.003939   
std       0.229235     0.022185     0.487321     0.302562     0.062641   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

       cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  \
count  8124.000000    8124.000000    8124.000000    8124.000000   
mean      0.450025       0.285574       0.000492       0.314623   
std       0.497527       0.451715       0.022185       0.464394   
min       0.000000       0.000000       0.000000       0.000000   
25%       0.000000       0.000000       0.000000       0.000000   
50%       0.000000       0.000000       0.000000       0.000000   
75%       1.000000       1.000000       0.000000       1.000000   
max       1.000000       1.000000       1.000000       1.000000   

       cap-surface_y  ...  population_s  population_v  population_y  \
count    8124.000000  ...   8124.000000   8124.000000   8124.000000   
mean        0.399311  ...      0.153619      0.497292      0.210734   
std         0.489787  ...      0.360605      0.500023      0.407855   
min         0.000000  ...      0.000000      0.000000      0.000000   
25%         0.000000  ...      0.000000      0.000000      0.000000   
50%         0.000000  ...      0.000000      0.000000      0.000000   
75%         1.000000  ...      0.000000      1.000000      0.000000   
max         1.000000  ...      1.000000      1.000000      1.000000   

         habitat_d    habitat_g    habitat_l    habitat_m    habitat_p  \
count  8124.000000  8124.000000  8124.000000  8124.000000  8124.000000   
mean      0.387494     0.264402     0.102413     0.035943     0.140817   
std       0.487208     0.441041     0.303209     0.186159     0.347854   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         habitat_u    habitat_w  
count  8124.000000  8124.000000  
mean      0.045298     0.023634  
std       0.207969     0.151914  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000     1.000000  

[8 rows x 117 columns]

In [51]:
# Creating model using Random Forest 
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier()
rf.fit(x_train,y_train)
y_pred=rf.predict(x_test)

In [52]:
print(y_pred)

['e' 'p' 'p' ... 'e' 'e' 'p']


In [53]:
from sklearn.metrics import confusion_matrix,accuracy_score
accuracy_score(y_test,y_pred)

1.0

In [54]:
confusion_matrix(y_test,y_pred)

array([[1040,    0],
       [   0,  991]], dtype=int64)

Random forest Classifier with accuracy 1.

In [55]:
#feature scaling
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train=sc.fit_transform(x_train)
x_test=sc.fit_transform(x_test)
x_train

array([[-0.23543673, -0.02219484, -0.78852564, ..., -0.40209177,
        -0.20942337, -0.16156443],
       [-0.23543673, -0.02219484,  1.26818957, ..., -0.40209177,
        -0.20942337, -0.16156443],
       [-0.23543673, -0.02219484, -0.78852564, ..., -0.40209177,
        -0.20942337, -0.16156443],
       ...,
       [-0.23543673, -0.02219484,  1.26818957, ...,  2.48699447,
        -0.20942337, -0.16156443],
       [-0.23543673, -0.02219484, -0.78852564, ...,  2.48699447,
        -0.20942337, -0.16156443],
       [-0.23543673, -0.02219484, -0.78852564, ..., -0.40209177,
        -0.20942337, -0.16156443]])

# Fitting Gaussian Naive Bayes to the Training

In [56]:
from sklearn.naive_bayes import GaussianNB
classifier=GaussianNB()
classifier.fit(x_train,y_train)

GaussianNB()

In [57]:
y_pred=classifier.predict(x_test)
y_pred

array(['e', 'p', 'p', ..., 'e', 'e', 'p'], dtype='<U1')

In [58]:
confusion_matrix(y_test,y_pred)

array([[1040,    0],
       [   5,  986]], dtype=int64)

In [59]:
accuracy_score(y_test,y_pred)

0.9975381585425899

 Gaussian Naive Bayes Classifier we can build a model with accuracy 0.9975381585425899